v23.1/transaction-retry-error-reference.md

```diff
@@ -1,54 +1,94 @@
+- If you're building an application with another driver or data access framework that is [supported by CockroachDB](third-party-database-tools.html), we recommend reusing the retry logic in our ["Simple CRUD" Example Apps](example-apps.html). For example, **Java** developers accessing the database with [JDBC](https://jdbc.postgresql.org) can reuse the example code implementing retry logic shown in [Build a Java app with CockroachDB](build-a-java-app-with-cockroachdb.html).
+- If you're building an application with a language and framework for which we do not provide example retry logic, you might need to write your own retry logic. For an example, see the [Client-side retry handling example](transaction-retry-error-example.html).
+- **Advanced users, such as library authors**: See [Advanced Client-Side Transaction Retries](advanced-client-side-transaction-retries.html).
+
+#### Client-side retry handling example
+
+For a conceptual example of application-defined retry logic, and testing that logic against your application's needs, see the [client-side retry handling example](transaction-retry-error-example.html).
+
+### Minimize transaction retry errors
 
-## Error reference
+In addition to the steps described in [Client-side retry handling](#client-side-retry-handling), which detail how to configure your application to restart a failed transaction, there are also a number of changes you can make to your application logic to prevent transaction retry errors from occurring in the first place:
+
+1. Design your applications to reduce network round trips by [sending statements in transactions as a single batch](transactions.html#batched-statements) (e.g., [common table expressions](common-table-expressions.html)), and by limiting the size of the result sets of your transactions so that CockroachDB is more likely to [automatically retry](transactions.html#automatic-retries) when previous reads are invalidated at a pushed timestamp.
+
+1. Break down larger transactions into smaller ones (e.g., [bulk deletes](bulk-delete-data.html)) wherever possible, and limit the number of affected rows by following [performance-tuning best practices](apply-statement-performance-rules.html) (e.g., query performance tuning, index design and maintenance, etc.). Not only will transactions run faster and hold locks for a shorter duration but also read-write transactions that affect many rows increases the likelihood of retry errors. For example, as the size of reads (number of rows) increases, the chances of read invalidation when the transaction’s timestamp is pushed due to a conflicting write increases. As the size of writes (number of rows written) increases, the chances of the transaction’s timestamp getting bumped by concurrent reads increases. 
```

Comment:
I was thinking we can separate this out into two bullets. Query tuning and Bulk Deletes. 

Although related, I do think the former is more common than the later and the context is a bit different.


Limit the number of affected rows by following [performance-tuning best practices](https://www.cockroachlabs.com/docs/stable/apply-statement-performance-rules.html) (e.g., query performance tuning, index design and maintenance, etc.). Reading less rows will decrease the likelihood of concurrent writes having their timestamp pushed (we should link out here again: https://www.cockroachlabs.com/docs/stable/architecture/transaction-layer.html#read-refreshing) which can lead to retry errors.

Break down larger transactions into smaller ones (e.g., [bulk deletes](https://www.cockroachlabs.com/docs/stable/bulk-delete-data.html)) to have transactions hold locks for a shorter duration. This will also decrease the likelihood of pushed timestamps and retry errors. For instance, as the size of writes (number of rows written) decreases, the chances of the (bulk delete) transaction’s timestamp getting bumped by concurrent reads decreases.